cmd/snap-confine/seccomp-support-test.c

```diff
@@ -0,0 +1,196 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "seccomp-support-ext.c"
+#include "seccomp-support.c"
+
+#include <glib.h>
+#include <glib/gstdio.h>
+
+static void make_seccomp_profile(struct sc_seccomp_file_header *hdr, int *fd,
+				 char **fname)
+{
+	*fd = g_file_open_tmp(NULL, fname, NULL);
+	g_assert_true(*fd > 0);
+	int written = write(*fd, hdr, sizeof(struct sc_seccomp_file_header));
+	g_assert_true(written == sizeof(struct sc_seccomp_file_header));
+}
+
+static void test_must_read_and_validate_header_from_file__happy(void)
+{
+	struct sc_seccomp_file_header hdr = {
+		.header[0] = 'S',
+		.header[1] = 'C',
+		.version = 1,
+	};
+	char SC_CLEANUP(sc_cleanup_string) * profile = NULL;
+	int SC_CLEANUP(sc_cleanup_close) fd = 0;
+	make_seccomp_profile(&hdr, &fd, &profile);
+
+	FILE *SC_CLEANUP(sc_cleanup_file) file =
+	    sc_must_read_and_validate_header_from_file(profile, &hdr);
+	g_assert_true(file != NULL);
+}
+
+static void test_must_read_and_validate_header_from_file__missing_file(void)
+{
+	struct sc_seccomp_file_header hdr;
+	const char *profile = "/path/to/missing/file";
+	const char *expected_err =
+	    "cannot open seccomp filter /path/to/missing/file: No such file or directory\n";
+
+	if (g_test_subprocess()) {
+		FILE *SC_CLEANUP(sc_cleanup_file) file =
+		    sc_must_read_and_validate_header_from_file(profile, &hdr);
+		// the function above is expected to call die()
+		g_assert_not_reached();
+		// reference "file" to keep the compiler from warning
+		// that "file" is unused
+		g_assert_null(file);
+	}
+	g_test_trap_subprocess(NULL, 0, 0);
+	g_test_trap_assert_failed();
+	g_test_trap_assert_stderr(expected_err);
+}
+
+static void must_read_and_validate_header_from_file_dies_with(struct
+							      sc_seccomp_file_header
+							      hdr, const char
+							      *err_msg)
+{
+	if (g_test_subprocess()) {
+		char SC_CLEANUP(sc_cleanup_string) * profile = NULL;
+		int SC_CLEANUP(sc_cleanup_close) fd = 0;
+		make_seccomp_profile(&hdr, &fd, &profile);
+
+		FILE *SC_CLEANUP(sc_cleanup_file) file =
+		    sc_must_read_and_validate_header_from_file(profile, &hdr);
+		// the function above is expected to call die()
+		g_assert_not_reached();
+		// reference "file" to keep the compiler from warning
+		// that "file" is unused
+		g_assert_null(file);
+	}
+
+	g_test_trap_subprocess(NULL, 0, 0);
+	g_test_trap_assert_failed();
+	g_test_trap_assert_stderr(err_msg);
+}
+
+static void test_must_read_and_validate_header_from_file__invalid_header(void)
+{
+	// workaround bug in gcc from 14.04, the pragma can be removed when
+	// we stop supporting 14.04
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmissing-braces"
+	struct sc_seccomp_file_header hdr = { 0 };
```

Comment:
You can also drop the whole initialization like that, drop the pragmas, and keep the test as is.

Reply:
2ae52cbfa2d8635f7365cbe8e21ad31dd4a02f57